File: rust/src/medium/two_buckets.rs
#[derive(PartialEq, Eq, Debug)]
pub enum Bucket {
    One,
    Two,
}

#[derive(PartialEq, Eq, Debug)]
pub struct BucketStats {
    pub moves: u8,
    pub goal_bucket: Bucket,
    pub other_bucket: u8,
}

fn pour_from_big_to_small(
    capacity_1: u8,
    capacity_2: u8,
    goal: u8,
    b1: u8,
    b2: u8,
    moves: u8,
) -> Option<BucketStats> {
    let moves = 99;
    let other_bucket = 0;

    Some(BucketStats {
        moves,
        goal_bucket: Bucket::Two,
        other_bucket,
    })
}

fn pour_from_small_to_big(
    capacity_1: u8,
    capacity_2: u8,
    goal: u8,
    b1: u8,
    b2: u8,
    moves: u8,
) -> Option<BucketStats> {
    if b1 == goal {
        return Some(BucketStats {
            moves,
            goal_bucket: Bucket::One,
            other_bucket: b2,
        });
    }

    if b2 == goal {
        return Some(BucketStats {
            moves,
            goal_bucket: Bucket::Two,
            other_bucket: b1,
        });
    }

    dbg!(b1, b2);
    match (b1, b2) {
        (0, b) => pour_from_small_to_big(capacity_1, capacity_2, goal, capacity_1, b, moves + 1),
        (a, 0) if capacity_2 == goal => {
            pour_from_small_to_big(capacity_1, capacity_2, goal, a, capacity_2, moves + 1)
        }
        (a, b) if a + b <= capacity_2 => {
            pour_from_small_to_big(capacity_1, capacity_2, goal, 0, a + b, moves + 1)
        }
        (a, b) if a + b - goal == capacity_2 => {
            pour_from_small_to_big(capacity_1, capacity_2, goal, goal, a + b - goal, moves + 1)
        }
        _ => None,
    }
}

pub fn solve(
    capacity_1: u8,
    capacity_2: u8,
    goal: u8,
    start_bucket: &Bucket,
) -> Option<BucketStats> {
    match start_bucket {
        Bucket::One => match capacity_1 <= capacity_2 {
            true => pour_from_small_to_big(capacity_1, capacity_2, goal, 0, 0, 0),
            false => pour_from_big_to_small(capacity_1, capacity_2, goal, 0, 0, 0),
        },
        Bucket::Two => match capacity_2 <= capacity_1 {
            true => pour_from_big_to_small(capacity_2, capacity_1, goal, 0, 0, 0),
            false => pour_from_big_to_small(capacity_2, capacity_1, goal, 0, 0, 0),
        },
    }
}

#[test]
fn goal_equal_to_other_bucket() {
    assert_eq!(
        solve(2, 3, 3, &Bucket::One),
        Some(BucketStats {
            moves: 2,

            goal_bucket: Bucket::Two,

            other_bucket: 2,
        })
    );
}

#[test]
fn test_case_1() {
    assert_eq!(
        solve(3, 5, 1, &Bucket::One),
        Some(BucketStats {
            moves: 4,

            goal_bucket: Bucket::One,

            other_bucket: 5,
        })
    );
}

#[test]
fn goal_equal_to_start_bucket() {
    assert_eq!(
        solve(1, 3, 3, &Bucket::Two),
        Some(BucketStats {
            moves: 1,

            goal_bucket: Bucket::Two,

            other_bucket: 0,
        })
    );
}

File: elixir/lib/medium/two_buckets.ex
defmodule TwoBuckets do
  defstruct [:bucket_one, :bucket_two, :moves]
  @type t :: %TwoBuckets{bucket_one: integer, bucket_two: integer, moves: integer}

  @doc """
  Find the quickest way to fill a bucket with some amount of water from two buckets of specific sizes.
  """
  @spec measure(
          size_one :: integer,
          size_two :: integer,
          goal :: integer,
          start_bucket :: :one | :two
        ) :: {:ok, TwoBuckets.t()} | {:error, :impossible}
  def measure(size_one, size_two, goal, :one) when size_one < size_two do
    pour_from_small_to_big(size_one, size_two, goal, size_one, 0, 1)
  end

  def measure(size_one, size_two, goal, :one) do
    pour_from_big_to_small(size_one, size_two, goal, size_one, 0, 1)
  end

  def measure(size_one, size_two, goal, :two) when size_two < size_one do
    pour_from_small_to_big(size_one, size_two, goal, 0, size_two, 1)
  end

  def measure(size_one, size_two, goal, :two) do
    pour_from_big_to_small(size_one, size_two, goal, 0, size_two, 1)
  end

  defp pour_from_big_to_small(_, _, goal, amount_one, amount_two, moves)
       when goal in [amount_one, amount_two] do
    {:ok, %TwoBuckets{bucket_one: amount_one, bucket_two: amount_two, moves: moves}}
  end

  defp pour_from_big_to_small(size_one, size_two, goal, amount_one, amount_two, moves) do
    case {amount_one, amount_two} do
      {0, b} when b > size_one ->
        pour_from_big_to_small(size_one, size_two, goal, size_one, b - size_one, moves + 1)

      {0, b} ->
        pour_from_big_to_small(size_one, size_two, goal, b, 0, moves + 1)

      {a, 0} ->
        pour_from_big_to_small(size_one, size_two, goal, a, size_two, moves + 1)

      {a, b} when a < size_one ->
        pour_from_big_to_small(size_one, size_two, goal, size_one, b - (size_one - a), moves + 1)

      {a, b} when a == size_one ->
        pour_from_big_to_small(size_one, size_two, goal, 0, b, moves + 1)

      _ ->
        {:error, :impossible}
    end
  end

  defp pour_from_small_to_big(_, _, goal, amount_one, amount_two, moves)
       when goal in [amount_one, amount_two] do
    {:ok, %TwoBuckets{bucket_one: amount_one, bucket_two: amount_two, moves: moves}}
  end

  defp pour_from_small_to_big(size_one, size_two, goal, amount_one, amount_two, moves) do
    case {amount_one, amount_two} do
      {^size_one, ^size_two} ->
        {:error, :impossible}

      {0, b} ->
        pour_from_small_to_big(size_one, size_two, goal, size_one, b, moves + 1)

      {a, 0} when size_two == goal ->
        pour_from_small_to_big(size_one, size_two, goal, a, size_two, moves + 1)

      {a, b} when a + b <= size_two ->
        pour_from_small_to_big(size_one, size_two, goal, 0, a + b, moves + 1)

      {a, b} when a + b - goal == size_two ->
        pour_from_small_to_big(size_one, size_two, goal, goal, a + b - goal, moves + 1)

      {a, b} when b < size_two ->
        pour_from_small_to_big(size_one, size_two, goal, a - (size_two - b), size_two, moves + 1)

      {a, b} when b == size_two ->
        pour_from_small_to_big(size_one, size_two, goal, a, 0, moves + 1)
    end
  end
end
